### ISSUE TITLE
Can the Windows installer actually "install" the DLL?

### ISSUE THREAD

**Issue by mgeier** [erikd/libsndfile#89]
I'm collaborating on [PySoundFile](http://github.com/bastibe/PySoundFile), a Python wrapper for libsndfile.

Currently, we have a copy of the current DLLs (32bit and 64bit) in our repository and we provide our own installers (generated by setuptools) which include the appropriate DLL and install it to some Python-specific directory.

We would like to get rid of the DLLs in our repository (and of our installers for different Windows and Python versions) and instead tell our users to download the official installer from your website.

However, when I tried your installer, it seemed that it merely extracts the contained files into a user-specifiable directory.

Would it be possible to create an installer that actually installs the appropriate DLL to an appropriate system directory on Windows?

In our concrete case we are using [CFFI](http://cffi.readthedocs.org/) to access the DLL (just to be clear: at runtime, not at installation time).
I tried (on Windows 7) to move the DLL to `C:\Windows\System32` and CFFI can find it automatically.
Alternatively, the directory containing the DLL could be added to the `PATH`, the CFFI would also find it.

I'm offering my help for changing the installer.
Are the scripts for creating the installer available somewhere?
I didn't find them in the repository ...

I think adding this to the official installer would be a win-win situation: we wouldn't have to create all the installers and you would benefit because all users would see your call for donations in your installer (or whatever you choose to put there).

**Comment by mgeier** [comment]
Any news on this?
How can I help?

**Comment by erikd** [comment]
It will be in the next release. I can't think of anything you can do right now to help. I'll let you know when I have a pre-release windows binary installer ready.

**Comment by mgeier** [comment]
That's great, thanks a lot!

Is there already an ETA for the next release?

**Comment by erikd** [comment]
There is no ETA. Since early December I have found and fixed 30+ input validation bugs. I *think* I am finally getting to the end of this process. Once I am I will turn my attention to a release.

**Comment by erikd** [comment]
Revisiting this.

Getting the INNO setup radio buttons working seems rather non-trivial. DO you have any experience with that?

**Comment by mgeier** [comment]
No, sorry, not at all.

If you want, I can give it a try, but I doubt that I have more luck than you ...

**Comment by mgeier** [comment]
@erikd, I just wanted to inform you on our policy change regarding the libsndfile installer:
Since release 0.7.0 (https://github.com/bastibe/PySoundFile/releases/tag/0.7.0), we are uploading Python "wheel" packages, which already include a DLL for Windows and a dylib for Mac OS X.
I cross-compiled the DLLs myself (using MXE) and uploaded them to a separate repository: https://github.com/bastibe/libsndfile-binaries

Long story short: we don't need your installer anymore, but thanks anyway!

I think the suggested changes still make sense in a more general context ...

**Comment by camlorn** [comment]
Just going to chime in with my two cents, as I was looking for something else and stumbled across this issue.
If you start copying dlls around on Windows, you can break things.  For me and everyone I know who does Windows development, the standard is to put the dlls alongside our apps as not to end up with version-related problems.  I don't personally know any developers (especially in Python) that put stuff in system32, etc.  I'm also pretty sure that MSDN suggests against that somewhere, nowadays.
I'm also doing Python wheels.  My suggested fix to this sort of thing is to run the installer, hard-code the paths, and attempt to grab the DLLs when you build the wheel.
If I was going to ask for anything, it would be this: vendor the dependencies you need for Windows in the Libsndfile repository, make the CMake generation scripts completely independent of bash in all ways, and leave the current installers alone.  Even if the installer is changed, the import libraries probably won't move, and you'll still  end up having to do this dance in your build system for all the languages that need those.  I'd kill for a vendorable Libsndfile, but that's probably a different issue all together.  I stumbled across this ticket because I'm working on setting up Windows CI this afternoon, and the installer is getting in the way.

**Comment by erikd** [comment]
Sorry, won't fix this.

**Comment by camlorn** [comment]
Can I give you a pull request or something?  This already has the CMake scripts so it's not like it's a new build system, and I really need to be able to vendor at some point in the future.  One of the goals of my current project is to be easy to build, and Libsndfile on Windows is one of the big things getting in the way.  I've been considering forking, but I really do not want to do that for all the obvious reasons.  If the issue here is you thinking I want to violate the LGPL or something, that's not it at all; it's simply that there's no standard install location for these things on Windows and telling people to either download two installers and run them or to copy these files to those places manually is a really sucky thing to have to do.

**Comment by erikd** [comment]
You can always submit a Pull Request, but a PR for what? Is that PR related to this issue of something else?

AFAIAC the CMake script is incomplete. If you going to improve that, then great!

You say "I really need to be able to vendor at some point in the future". Sorry, I don't understand what that means.

I open to fixing the Windows install issue, but I didn't think the solution proposed here was a good one. Also since I am not a Windows user, I have little motivation to fix it.

**Comment by camlorn** [comment]
On Windows and for an easy experience for building, it's easiest to just hook in the things you depend on.  Ideally they then pick up your build settings, etc.  Program files is not where usually libraries live, and as it stands you have to run two installers and pull out files to use it with your app properly.
If CMake works and Libsndfile is edited to be in line with VS2015 (which is better about C99) then anyone can just download and build it, and anyone using CMake can just include it directly in their project.  I and others I know call such inclusion vendoring, though some people go so far as to stipulate that you need to rewrite the build system.  Firefox does this for a lot of their components.
But you also gain the CMake packaging stuff, which lets you build installers for your app that include all the files in the right places.
But you're right; we may be beyond this ticket.  I assumed that this was the appropriate place, because it was discussion of installation and the suggested solution would actually be really bad for  anything I do in future.  If you want me to open a separate issue for discussion of improvements to the CMake stuff and possible editing to work with at least VS2015, I can.  I don't know when I'll get to it, but it's certainly on my to-do list one way or another.
